### PR TITLE
add agentAuth client function

### PR DESCRIPTION
**Pull request by yij793**
nanoside: https://github.com/nanoPayinc/NANOPAY/pull/11941

### CODE REVIEW

File: src/foam/nanos/auth/AgentAuthClient.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  methods: [
+    async function actAs(x, business) {
+      let result = await this.delegate.actAs(x, business);
+      if ( result ) {
+        await x.fetchGroup();
+        x.subject.user = business;
+        x.crunchController.purgeCachedCapabilityDAOs();
+        x.client.menuDAO.cmd_(x, foam.dao.CachingDAO.PURGE);
+        x.client.menuDAO.cmd_(x, foam.dao.AbstractDAO.RESET_CMD);
+        x.crunchController.purgeCachedCapabilityDAOs();
```
Comment (zahnno) [inline src/foam/nanos/auth/AgentAuthClient.js]: looks like a duplicate of ln 20

File: src/foam/nanos/auth/AgentAuthClient.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  extends: 'foam.nanos.auth.ProxyAuthService',
+  documentation: `
+  `,
+
+  methods: [
+    async function actAs(x, business) {
+      let result = await this.delegate.actAs(x, business);
+      if ( result ) {
+        await x.fetchGroup();
+        x.subject.user = business;
```
Comment (zahnno) [inline src/foam/nanos/auth/AgentAuthClient.js]: I believe we'll want to be setting the subject realUser here to the result of actAs as well

File: src/foam/nanos/auth/AgentAuthClient.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth',
+  name: 'AgentAuthClient',
+  extends: 'foam.nanos.auth.ProxyAuthService',
+  documentation: `
+  `,
```
Comment (zahnno) [inline src/foam/nanos/auth/AgentAuthClient.js:14]: Can you provide documentation on what the client method decorator is doing

File: src/foam/nanos/auth/AgentAuthClient.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    async function actAs(x, business) {
+      let result = await this.delegate.actAs(x, business);
+      if ( result ) {
+        await x.fetchGroup();
+        x.subject.user = business;
+        x.crunchController.purgeCachedCapabilityDAOs();
+        x.client.menuDAO.cmd_(x, foam.dao.CachingDAO.PURGE);
+        x.client.menuDAO.cmd_(x, foam.dao.AbstractDAO.RESET_CMD);
+        x.crunchController.purgeCachedCapabilityDAOs();
+        x.initLayout.resolve();
```
Comment (zahnno) [inline src/foam/nanos/auth/AgentAuthClient.js]: You can import all of the methods called on x

**Comment by zahnno** [comment]
Why not add this to the agentAuth service client?

**Review by zahnno — APPROVED** [review]

File: src/foam/nanos/auth/AgentAuthClient.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    async function actAs(x, business) {
+      let result = await this.delegate.actAs(x, business);
+      if ( result ) {
+        await x.fetchGroup();
+        this.subject.user = business;
+        this.subject.realUser = result;
+        this.menuDAO.cmd_(x, foam.dao.CachingDAO.PURGE);
+        this.menuDAO.cmd_(x, foam.dao.AbstractDAO.RESET_CMD);
+        this.crunchController.purgeCachedCapabilityDAOs();
+        this.initLayout.resolve();
```
Comment (zahnno) [inline src/foam/nanos/auth/AgentAuthClient.js]: This is not available in foam.

File: src/foam/nanos/auth/AgentAuthClient.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    async function actAs(x, business) {
+      let result = await this.delegate.actAs(x, business);
+      if ( result ) {
+        await x.fetchGroup();
+        this.subject.user = business;
+        this.subject.realUser = result;
+        this.menuDAO.cmd_(x, foam.dao.CachingDAO.PURGE);
+        this.menuDAO.cmd_(x, foam.dao.AbstractDAO.RESET_CMD);
+        this.crunchController.purgeCachedCapabilityDAOs();
+        return;
```
Comment (kgrgreer) [inline src/foam/nanos/auth/AgentAuthClient.js]: return statement not needed

File: src/foam/nanos/auth/AgentAuthClient.js (hunk 3 above)
Comment (kgrgreer) [reply src/foam/nanos/auth/AgentAuthClient.js:14]: It seems like the main purpose is to purge caches when the user calls auth.actAs(). You should mention that in the documentation.

**Pull request merged**